src/structure/spatial.jl

```diff
@@ -1,66 +1,298 @@
+    "N"  in atomnames(res)      ? nothing : throw(ArgumentError("Atom with atom name \"N\" not found in residue"))
+    "CA" in atomnames(res)      ? nothing : throw(ArgumentError("Atom with atom name \"CA\" not found in residue"))
+    return dihedralangle(res_prev["CA"], res_prev["C"], res["N"], res["CA"])
+end
 
-"Get the minimum distance between two `StructuralElementOrList`s."
-function distance(element_one::StructuralElementOrList, element_two::StructuralElementOrList, selector_functions::Function...)
-    coords_one = coordarray(element_one, selector_functions...)
-    coords_two = coordarray(element_two, selector_functions...)
-    count_one = size(coords_one, 2)
-    count_two = size(coords_two, 2)
-    sq_dists = zeros(count_one, count_two)
-    for i in 1:count_one, j in 1:count_two
-        @inbounds sq_dists[i,j] = (coords_one[1,i] - coords_two[1,j]) ^ 2 + (coords_one[2,i] - coords_two[2,j]) ^ 2 + (coords_one[3,i] - coords_two[3,j]) ^ 2
+"""
+Calculate the phi angle in radians for an `AbstractResidue`. Arguments are the
+residue (atoms "N", "CA" and "C" required) and the previous residue (atom "C"
+required).
+"""
+function phiangle(res::AbstractResidue, res_prev::AbstractResidue)
+    "C"  in atomnames(res_prev) ? nothing : throw(ArgumentError("Atom with atom name \"C\" not found in previous residue"))
+    "N"  in atomnames(res)      ? nothing : throw(ArgumentError("Atom with atom name \"N\" not found in residue"))
+    "CA" in atomnames(res)      ? nothing : throw(ArgumentError("Atom with atom name \"CA\" not found in residue"))
+    "C"  in atomnames(res)      ? nothing : throw(ArgumentError("Atom with atom name \"C\" not found in residue"))
+    return dihedralangle(res_prev["C"], res["N"], res["CA"], res["C"])
+end
+
+"""
+Calculate the psi angle in radians for an `AbstractResidue`. Arguments are the
+residue (atoms "N", "CA" and "C" required) and the next residue (atom "N"
+required).
+"""
+function psiangle(res::AbstractResidue, res_next::AbstractResidue)
+    "N"  in atomnames(res)      ? nothing : throw(ArgumentError("Atom with atom name \"N\" not found in residue"))
+    "CA" in atomnames(res)      ? nothing : throw(ArgumentError("Atom with atom name \"CA\" not found in residue"))
+    "C"  in atomnames(res)      ? nothing : throw(ArgumentError("Atom with atom name \"C\" not found in residue"))
+    "N"  in atomnames(res_next) ? nothing : throw(ArgumentError("Atom with atom name \"N\" not found in next residue"))
+    return dihedralangle(res["N"], res["CA"], res["C"], res_next["N"])
+end
+
+
+"""
+Calculate the `Vector`s of phi and psi angles of a `StructuralElementOrList`.
+The vectors have `nothing` for residues where an angle cannot be calculated,
+e.g. due to missing atoms or lack of an adjacent residue.
+Additional arguments are residue selector functions - only residues that return
+`true` from the functions are retained.
+"""
+function ramachandranangles(el::StructuralElementOrList,
```

Comment:
Including `nothing` in an array will result in very bad performance in Julia. I think using `NaN`, which is a value of `Float64`, or using `NullableArray` of NullableArrays.jl is a better way to avoid the problem.